src/main/java/org/jenkinsci/plugins/workflow/log/FileLogStorage.java

```diff
@@ -0,0 +1,276 @@
+                        long pos = start;
+                        boolean hadLastId;
+                        @Override public void write(int b) throws IOException {
+                            while (lastTransition < pos && !eof) {
+                                String line = indexBR.readLine();
+                                if (line == null) {
+                                    eof = true;
+                                    break;
+                                }
+                                int space = line.indexOf(' ');
+                                try {
+                                    lastTransition = Long.parseLong(space == -1 ? line : line.substring(0, space));
+                                } catch (NumberFormatException x) {
+                                    LOGGER.warning("Ignoring corrupt index file " + index);
+                                }
+                                lastId = space == -1 ? null : line.substring(space + 1);
+                            }
+                            if (pos == lastTransition) {
+                                if (hadLastId) {
+                                    w.write(LogStorage.endStep());
+                                }
+                                hadLastId = lastId != null;
+                                if (lastId != null) {
+                                    w.write(LogStorage.startStep(lastId));
+                                }
+                            }
+                            super.write(b);
+                            pos++;
+                        }
+                        @Override public void flush() throws IOException {
+                            if (lastId != null) {
+                                w.write(LogStorage.endStep());
+                            }
+                            super.flush();
+                        }
+                    });
+                    ConsoleAnnotators.setAnnotator(caos.getConsoleAnnotator());
+                    return r;
+                }
+            }
+        };
+    }
+
+    @Override public AnnotatedLargeText<FlowNode> stepLog(FlowNode node, boolean complete) {
+        String id = node.getId();
+        try (ByteBuffer buf = new ByteBuffer();
+             RandomAccessFile raf = new RandomAccessFile(log, "r");
+             BufferedReader indexBR = index.isFile() ? Files.newBufferedReader(index.toPath(), StandardCharsets.UTF_8) : new BufferedReader(new NullReader(0))) {
+            String line;
+            long pos = -1; // -1 if not currently in this node, start position if we are
+            while ((line = indexBR.readLine()) != null) {
+                int space = line.indexOf(' ');
+                long lastTransition = -1;
+                try {
+                    lastTransition = Long.parseLong(space == -1 ? line : line.substring(0, space));
+                } catch (NumberFormatException x) {
+                    break; // corrupt index file; forget it
```

Comment:
🐛  One line that is not number-parseable (perhaps due to interrupted writes or OS write caching under circumstances similar to those described in https://github.com/jenkinsci/jenkins/pull/3170) will prevent reading the entire log file IIUC.  What you *probably* want to do is log the exception somewhere and `continue`.

Reply:
Though who knows, this logic is like reading Don Quixote in the original old Spanish

Reply:
Actually my uncle-in-law is a Don Quixote expert!

> will prevent reading the entire log file

Will prevent the step log association. The whole build log will be fine except for possibly missing some step annotations.

> and `continue`

Not so simple. If `log-index` is corrupt in one line, at best you can hope that the _next_ line (next transition) is accurate. Certainly this step’s log is going to be incomplete or empty.

Reply:
@jglick The catch here is that we may have a single log index line that is corrupted but the rest may be okay (for example if the AWS or K8S volume is abruptly yanked) but the master restarts and resumes the build later.  

So it is quite possible that the index may have a *single* corrupted or missing entry with the rest of the index being fine. 

Because we break at the *first* NumberFormatException, we will not see logs for any FlowNodes after the failure.  However if we do a `continue` we can accept that failure and still get log context for the following steps.

Reply:
Resolved via code change.

Reply:
@jglick Side comment: the simile of reading Don Quixote in the original Spanish is particularly appropriate because I did a fair bit of that (laboriously with painful effort because while I was almost-fluent in the language at the time, the particular usages were quite painful).